sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala

```diff
@@ -64,31 +64,54 @@ object ResolveHints {
             _.toUpperCase(Locale.ROOT)).contains(hintName.toUpperCase(Locale.ROOT))))
     }
 
+    // This method checks if given multi-part identifiers are matched with each other.
+    // The [[ResolveJoinStrategyHints]] rule is applied before the resolution batch
+    // in the analyzer and we cannot semantically compare them at this stage.
+    // Therefore, we follow a simple rule; they match if an identifier in a hint
+    // is a tail of an identifier in a relation. This process is independent of a session
+    // catalog (`currentDb` in [[SessionCatalog]]) and it just compares them literally.
+    //
+    // For example,
+    //  * in a query `SELECT /*+ BROADCAST(t) */ * FROM db1.t JOIN t`,
+    //    the broadcast hint will match both tables, `db1.t` and `t`.
+    //  * in a query `SELECT /*+ BROADCAST(default.t) */ * FROM default.t JOIN t`,
+    //    the broadcast hint will match the left-side table only, `default.t`.
+    private def matchedIdentifier(identInHint: Seq[String], identInQuery: Seq[String]): Boolean = {
+      if (identInHint.length <= identInQuery.length) {
+        identInHint.zip(identInQuery.takeRight(identInHint.length))
+          .forall { case (i1, i2) => resolver(i1, i2) }
+      } else {
+        false
+      }
+    }
+
     private def applyJoinStrategyHint(
         plan: LogicalPlan,
-        relations: mutable.HashSet[String],
+        relationsInHint: Seq[Seq[String]],
+        appliedRelations: mutable.ArrayBuffer[Seq[String]],
         hintName: String): LogicalPlan = {
       // Whether to continue recursing down the tree
       var recurse = true
 
       val newNode = CurrentOrigin.withOrigin(plan.origin) {
         plan match {
           case ResolvedHint(u @ UnresolvedRelation(ident), hint)
-              if relations.exists(resolver(_, ident.last)) =>
-            relations.remove(ident.last)
```

Comment:
Seems like the relation name in hint can only match one relation before? As we remove it from the set after finding one match.

Reply:
Ur, nice catch. I didn't notice this new behaivour in v3.0. We should keep this beheivour?
```
scala> sql("CREATE DATABASE db1")
scala> sql("CREATE TABLE db1.t(key int)")
scala> sql("CREATE TABLE t(key int)")
scala> sql("""SELECT /*+ MAPJOIN(t) */ * FROM db1.t JOIN t""").explain(true)
== Parsed Logical Plan ==
'UnresolvedHint MAPJOIN, ['t]
+- 'Project [*]
   +- 'Join Inner
      :- 'UnresolvedRelation [db1, t]
      +- 'UnresolvedRelation [t]

// v3.0
== Analyzed Logical Plan ==
key: int, key: int
Project [key#0, key#1]
+- Join Inner
   :- ResolvedHint (strategy=broadcast)
   :  +- SubqueryAlias `db1`.`t`
   :     +- Relation[key#0] parquet
   +- SubqueryAlias `default`.`t`
      +- Relation[key#1] parquet

// v2.4
== Analyzed Logical Plan ==
key: int, key: int
Project [key#20, key#21]
+- Join Inner
   :- ResolvedHint (broadcast)
   :  +- SubqueryAlias `db1`.`t`
   :     +- HiveTableRelation `db1`.`t`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [key#20]
   +- ResolvedHint (broadcast)
      +- SubqueryAlias `default`.`t`
         +- HiveTableRelation `default`.`t`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [key#21]
```

Reply:
But, the v3.0 behaivour looks inconsitent;
```

scala> sql("SET spark.sql.caseSensitive=false")
scala> sql("""SELECT /*+ MAPJOIN(T) */ * FROM db1.t JOIN t""").explain(true)
20/03/19 15:55:48 WARN HintErrorLogger: Count not find relation 'T' specified in hint 'MAPJOIN(T)'.
// v3.0
== Parsed Logical Plan ==
'UnresolvedHint MAPJOIN, ['T]
+- 'Project [*]
   +- 'Join Inner
      :- 'UnresolvedRelation [db1, t]
      +- 'UnresolvedRelation [t]

== Analyzed Logical Plan ==
key: int, key: int
Project [key#10, key#11]
+- Join Inner
   :- ResolvedHint (strategy=broadcast)
   :  +- SubqueryAlias `db1`.`t`
   :     +- Relation[key#10] parquet
   +- ResolvedHint (strategy=broadcast)
      +- SubqueryAlias `default`.`t`
         +- Relation[key#11] parquet
```
The root cause of this behaivour above is that `relations.remove(ident.last)` is case-sensitive.

Reply:
I'm OK to add hint to all the matches. It's better than 2.4 which kinds of randomly picked one match according to the plan traverse order.

Reply:
ok

Reply:
+1